modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CustomViewsControls.js

```diff
@@ -57,8 +64,52 @@ const CustomViewsControls = () => {
 				},
 				{
 					label: Liferay.Language.get('save'),
-					onClick: ({processClose}) => {
-						processClose();
+					onClick: () => {
+						const url = new URL(`${appURL}/fds/${id}/custom-views`);
+
+						const data = {
+							name: customViewNameInputRef.current.value,
+							plid: themeDisplay.getPlid(),
+							portletId,
+							viewState: {
+								contentRenderer: activeView.contentRenderer,
+							},
```

Comment:
You made very interesting points, let me try to keep this short.

> If I am understanding the first-class citizens approach, the idea is that a custom view is a comprehensive definition of FDS frontend.

We need to turn the concept of "view" into something FDS can not only "apply" but also generate. This feature makes us to set a new architectural pattern concerning the way FDS produces, understands and provisions "views". 

Central point here is how a "view" (quoted) is represented for the FDS. As you pointed out, elements making up what we call "view" come from different properties we feed into the (react) component via taglib machinery, including `views` prop but also `filters`, `sorting` and others which depend on the particular visualization method, such as column widths (for tables). This is not necessarily bad, it just means that the props needed to represent a "view" are not under the same umbrella.

Back to the arch pattern, there are some clear separation of concerns that must be kept:
* View definition model: "view" (quoted) shall be understood by the FDS as a flexible specification of how to behave when rendering itself, 
  * Openness: It must account for defaults/optional values and also be prepared to grow in the future, as we support more attributes that can be part of a view definition. This way, we are not forcing view providers to be "up to date" when it comes to generate views, yet support new features as needed.
  * Unified: the concept does not depend on who creates the view. User-created views (traditionally called "custom views"), views on custom objects created by admins via object view builder (called "system views"), and programmatically created views for commerce tables are the same thing to the FDS eyes.
  * Self-contained: as explained above, there should be no dependency with external data for the FDS to render a view
* Multiple view generators: to produce a view we have the serializer in the backend, which is fed by a set of java classes. That's how the taglib feeds the react component. Now, we need the react component to be able to generate views too. This in fact creates a new provider for views.
* View injection point: currently, taglib injects necessary view info to the component at render time, and that is a fixed set. We can extend this concept to make FDS to retrieve additional view information w/o the need to reload the page, and include those new views in the selectable options offered to the user. I am not sure we need this at this moment, but that's an option.
* View persistence: until very recently, views were "persisted" in the JVM as classes ready to use for the serializer. Object feature offers a DB-based view persistence which is "transformed" by providing certain classes that [act as bridge](https://github.com/liferay/liferay-portal/blob/57bf3f7c2943ebc2cb268b4a131d1ca702cc2f2b/modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/entries/frontend/data/set/view/table/ObjectEntriesTableFDSView.java#L51) between the interfaces expected by the FDS view serializer and the DB persistence. The more "expresiveness" we want to support for the views, the more "complex" that DB-persistence and corresponding Java interfaces will become.
* Favourites and other context-info: this is a totally separated information not related to a view itself but rather a connection between views, users and FDS instances. These have to be persisted elsewhere, not inside the views.

With all this in mind, when FDS becomes a view generator, the way to persist those views sounds like a crucial decision. We can open an endpoint to receive the "view" JSON representation, which includes a lot of properties. But, where that information should go?

(I'll leave the comment at this point, to discuss the options later on)

Reply:
Thanks for clarifications! I'll continue with the quotes convention, to distinguish concepts from code object. 

I can see the logic in the principle of a "view" containing entire rendering specification. It is simple and clean. Technically, it is no problem to add contents of active `view`, `filter`, `sorting` etc to `viewState`.

However, there are some potential pitfalls with that approach. It comes from the fact that not all customizations are made equal. There are two broad groups of "view" properties:
- a set of mandatory, _unchangable_ properties. These include:
  - set of all available fields
  - set of all available filters
  - set of all available sorts
- a set of changeable properties. These include:
  - set of displayed fields
  - items layout and visualizations (table, cards)
  - data renderers for fields
  - UX features like pagination
  - set of active filters
  - set of active sorts

Changeable properties are easy, we are on the same page on them. They can be generated by anything and be the same thing to the FDS eyes. The problem are unchangeable properties. FDS can never generate a new field. It needs to be added in DB first, either through service builder or UI (as in objects). FDS can never generate a new type of filter. It needs to be created in code first, as a React component. 

So, here is the potential pitfall. If FDS view generation creates a comprehensive "view" definition, and that "view" definition is detached from any dependency on external data, that means that we are stuck with unchangeable properties. "view" will never be able to reflect changes in fields or filters that developers would wish to make, like additions, modifications or deletions. Are we ok with this? To apply changes, "view" may need to be deleted and recreated. We may need some UI that informs users that changes were made.

A solution, that I am suggesting, is to have "view definition" as a separate concept and implementation from "view state". "view definition" would be something that provides defaults. Something that reflects all available options in a "view". In technical terms, this would be an aggregate of `filters` prop, `sorting` prop, and a new special comprehensive `view`, containing all fields. With this, we could offer a 'smart' merge of saved and new options in UI.

Reply:
I've been struggling with this for months :) Initially, I also thought about making a distinction between the "definition" and the "state" of a view. And indeed I had the idea to move forward with the approach. Two separate facts prevented me to keep pushing the concept (as I thought it for the first time):
- We added "prepopulated" filters into the view definition, kinda shrinking the space for the concept of "state", merging it into the definition itself
- Problem of view evolution is there, no matter whether they get created by FDS or by other generator (Java classes/object view builder)

The fact that we make views self contained does not prevent us to distinguish between the "unchangeable" and "changeable" parts of it, though. This is important when dealing with evolvability.

So yes, this is an issue related with evolvability. The full set of fields and filters (what you called "unchangeable") may change over time. Just consider an object with a view, and then we add another field to the object definition. Here, view has not changed at all, there's just a new field which is unmapped to the view columns. If we stop there, there'd not be any issue FDS-wise. But then we go ahead and edit the view, mapping the new field in it. This effectively creates a new "version" of that view. (this is why I said the problem is there before FDS becomes able to generate views). This system view would then be populated into the frontend when users browse the page. So far, so good, because view was changed.

When users go to a FDS showing those objects, the "default" view will then include the new field, and maybe, a new filter for it. However, the existing custom views created by users "from" the previous version of the default view won't include neither the filter nor the field mapping. So at this point, users may wonder what to do. Should I re-create my custom view manually using the FDS? Or should the component be smart enough to merge the new info? 

In the latter case, there are cases where we can "do something", bascially, concerning the _unchangeable_ stuff:
* Component can add the new filter for that field, making it available for usage when the custom view is active
* Component can add the new field to the list of available fields to show/hide, when the custom view is active

But also, there are things that component can't solve, suspiciously related to the _changeable_ stuff:
* How component knows whether the new field should be ever shown/hidden in that custom view? → we need defaults
* How component knows the column width the  new field should take if shown? → we need defaults
* How component knows any sorting/filtering for that field? → we need defaults (no filter/no sorting)

We can make this exercise in case we change the type of a field or we remove a field in the object definition. For these cases, the "system" view should change immediately as per object feature integrity, so that when shipped to the frontend, it includes the changes. 

So, we could assume that "data evolution implies that views should evolve accordingly".

In the case of object view builder and the ones built via code, this is a reasonable expectation, because:
- The guy in charge of evolving data model also cares of evolving the view
- The system may do some work at keeping view definitions consistent with data

However, views generated from the FDS don't have that context because the guy evolving the data does not care about custom views. And also the guy creating the views (end user) is not likely to be the one modelling the data.

So, even if we have to put some sanity checks on views so that FDS can fail gracefully, there's a limit on how smart it can be when underlying data changes but views do not.

At this point, I can think on 2 potential ways to move forward:

1. Persist custom views as object views too, under the assumption that object feature will make updates on all views for a given object when its definition changes. Basically, we trust "The system may do some work at keeping view definitions consistent with data" 
2. Reference the "source" view from the custom view, i.e. the system view which was active when custom view was saved. Even if they're self contained, we can always compare what you called "unchangeable" properties in both views and update them into the custom view.

From the point of view of a "agnostic" component, option 1 sounds cleaner but intriduces a dependency with an end-user feature we don't control. We can "hide" it under the JAXRS endpoint implementation.

Option 2 supports different places to persist the view but requires some processing, perhaps in the backend service in charge of providing the views to the frontend.

Hope we're on the same page after such a long comment ;)

Reply:
I see you gave a lot of thought into exactly how the "smart" merge could happen :) I think it's overall doable. At some debatable points, like reasonable expectation of users, we can be either opinionated, or offer API to choose behavior.

But the key point I'm making is that none of this is possible if we have self-contained "views". There must be a "default view" / "view definition", something that we can compare with the saved "view". That means that the saved "view" must contain a reference to it, so it is not self-contained.